### PR TITLE
Add basic support for notifications to sftpgws

### PR DESCRIPTION
**Pull request by mic-e**

### CODE REVIEW

File: abrechnung/sftpgws.py [hunk 1]
```diff
@@ -91,17 +82,34 @@ async def run(self):
...
+        }) + "\n"
+
+        if payload_json['connections'] == '*':
+            tx_queues = self.tx_queues.values()
+        else:
+            tx_queues = (
+                self.tx_queues[connection_id]
+                for connection_id in payload_json['connections']
+                if connection_id in self.tx_queues
+            )
```
Comment (TheJJ) [inline abrechnung/sftpgws.py]: better use a real loop and fill `tx_queues` and also check if the queue is actually known (else skip it)

File: abrechnung/sftpgws.py [hunk 2]
```diff
@@ -91,17 +82,34 @@ async def run(self):
...
 
-        # TODO notify relay to correct client
-
-        raise NotImplementedError()
+        for tx_queue in tx_queues:
+            tx_queue.put_nowait(message)
```
Comment (TheJJ) [inline abrechnung/sftpgws.py]: if one queue is full, this excepts. other queues should still be filled.

File: abrechnung/sftpgws.py [hunk 3]
```diff
@@ -114,35 +122,45 @@ async def handle_ws_connection(self, request):
...
+                "select * from client_connected($1);",
                 self.cfg['websocket']['id']
             )
 
-            self.open_connections[client_id] = ws
+            # when too many unsent messages are in the queue, error.
+            tx_queue = asyncio.Queue(maxsize=1000)
+
+            self.tx_queues[connection_id] = tx_queue
+            asyncio.create_task(self.tx_task(ws, tx_queue))
```
Comment (TheJJ) [inline abrechnung/sftpgws.py]: we should track this task somehow, and we should add a future-done-callback to somehow handle exceptions in the task (e.g. log them)

File: abrechnung/sftpgws.py [hunk 4]
```diff
@@ -114,35 +122,45 @@ async def handle_ws_connection(self, request):
...
-                        await ws.send_str("".join((response, "\n")))
-
+                        response = (await self.ws_message(connection, connection_id, msg.data)) + '\n'
                     except Exception as exc:
                         import traceback
                         traceback.print_exc()
-                        await ws.send_str(f"error in ws_message: {exc}\n")
-
+                        response = f'error in ws_message: {exc}\n'
+                    tx_queue.put_nowait(response)
```
Comment (TheJJ) [inline abrechnung/sftpgws.py]: i think the client would love it iwe we always returned json

File: abrechnung/sftpgws.py [hunk 5]
```diff
@@ -166,7 +186,13 @@ async def ws_message(self, connection, client_id, msg_raw):
 
             # construct the sql query
             # SELECT $function($arg0, $arg1, ...);
```
Comment (TheJJ) [inline abrechnung/sftpgws.py]: can also be call then

**Review by TheJJ — CHANGES_REQUESTED** [review]
yes yes

File: abrechnung/sftpgws.py (hunk 1 above)
Comment (mic-e) [reply abrechnung/sftpgws.py]: queue is known?

Comment (TheJJ) [reply abrechnung/sftpgws.py]: ah right sli

**Pull request merged**